### ISSUE TITLE
[BUG]: Kubeflow Pipelines GCP example doesn't work. Kubeflow metadata store is not ready yet

### ISSUE THREAD

**Issue by VictorW96** [zenml-io/zenml#728]
### Contact Details [Optional]

victor.wolf@qaware.de

### System Information

ZenML version: 0.9.0
Install path: /home/victor/.local/lib/python3.8/site-packages/zenml
Python version: 3.8.10
Platform information: {'os': 'linux', 'linux_distro': 'ubuntu', 'linux_distro_like': 'debian', 'linux_distro_version': '20.04'}
Environment: native
Integrations: ['gcp', 'kubeflow', 'mlflow', 'scipy', 'seldon', 'sklearn', 'tensorflow']

### What happened?

The kubeflow metadata store seems to be not reachable. If I list the metadata stores the output is as follows:
![grafik](https://user-images.githubusercontent.com/40980071/175819729-bff90c2f-dd5d-4f86-89eb-55b713f4f2c3.png)
I find it curious that the IP address points to localhost. 
I also saw that kubeflow changed its metadata store implementation https://www.kubeflow.org/docs/components/pipelines/concepts/metadata/ . Is this a problem? 

My GCP kubeflow configuration looks like this:
![grafik](https://user-images.githubusercontent.com/40980071/175819921-6524a2b2-a3c3-4fb4-b625-df2c57f17974.png)

### Reproduction steps

Follow the kubeflow pipeline orchestration example. In particular follow the step: **Run the same pipeline on Kubeflow Pipelines deployed to GCP**

### Relevant log output

```shell
Provisioning resources for stack 'gcp_kubeflow_stack'.
Provisioning local Kubeflow Pipelines deployment...
Provisioned resources for KubeflowMetadataStore(type=metadata_store, flavor=kubeflow, name=kubeflow_metadata_store, uuid=165eb418-8d7b-42b9-a2a9-29d01259486d, upgrade_migration_enabled=False, host=127.0.0.1, port=8081).
Resuming provisioned resources for stack gcp_kubeflow_stack.
Started Kubeflow Pipelines UI daemon (check the daemon logs at /home/victor/.config/zenml/kubeflow/eda1b333-39a8-4a3b-9647-0f00fb78dab2/kubeflow_daemon.log in case you're not able to view the UI). The Kubeflow Pipelines UI should now be accessible at http://localhost:8080/.
Resumed resources for KubeflowOrchestrator(type=orchestrator, flavor=kubeflow, name=gcp_kubeflow_orchestrator, uuid=eda1b333-39a8-4a3b-9647-0f00fb78dab2, custom_docker_base_image_name=None, kubeflow_pipelines_ui_port=8080, kubeflow_hostname=None, kubernetes_context=gke_ai-gilde_europe-west4-a_cluster-1, synchro
nous=False, skip_local_validations=False, skip_cluster_provisioning=False, skip_ui_daemon_provisioning=False).
Started Kubeflow Pipelines Metadata daemon (check the daemonlogs at /home/victor/.config/zenml/kubeflow/eda1b333-39a8-4a3b-9647-0f00fb78dab2/metadata-store/165eb418-8d7b-42b9-a2a9-29d01259486d/kubeflow_daemon.log in case you're not able to access the pipelinemetadata).
Waiting for the Kubeflow metadata store to be ready (this might take a few minutes).
The Kubeflow metadata store is not ready yet. Waiting for 10 seconds...
The Kubeflow metadata store is not ready yet. Waiting for 10 seconds...
The Kubeflow metadata store is not ready yet. Waiting for 10 seconds...
The Kubeflow metadata store is not ready yet. Waiting for 10 seconds...
The Kubeflow metadata store is not ready yet. Waiting for 10 seconds...
The Kubeflow metadata store is not ready yet. Waiting for 10 seconds...
The Kubeflow metadata store is not ready yet. Waiting for 10 seconds...
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by htahir1** [comment]
Can you try doing `zenml stack up` again and see if it solves the problem?

**Comment by VictorW96** [comment]
No the same thing happens again. I think the metadata store and the kubeflow orchestrator are both thought of local by zenml. When I deprovision the resources with `zenml stack down --force
` I get the following output:
`Deprovisioning resources for stack 'gcp_kubeflow_stack'.
Deprovisioned resources for KubeflowOrchestrator(type=orchestrator, flavor=kubeflow, name=gcp_kubeflow_orchestrator, uuid=eda1b333-39a8-4a3b-9647-0f00fb78dab2, custom_docker_base_image_name=None, kubeflow_pipelines_ui_port=8080, kubeflow_hostname=None, kubernetes_context=gke_ai-gilde_europe-west4-a_cluster-1, s
ynchronous=False, skip_local_validations=False, skip_cluster_provisioning=False, skip_ui_daemon_provisioning=False).
Local kubeflow pipelines deployment deprovisioned.
Deprovisioned resources for KubeflowMetadataStore(type=metadata_store, flavor=kubeflow, name=kubeflow_metadata_store, uuid=165eb418-8d7b-42b9-a2a9-29d01259486d, upgrade_migration_enabled=False, host=127.0.0.1, port=8081).`

How do I tell ZenML to explicitly use the gcloud kubeflow?

**Comment by stefannica** [comment]
Hi Victor, thank you for opening this issue. Your stack setup looks good. Even if you see `host=127.0.0.1` in your metadata store configuration, that is expected. When you run `zenml stack up`, the gRPC metadata service port is forwarded locally via a `kubectl port-forward` command, which is why you see ZenML trying to access the metadata store on your localhost.

Here are a few things you can try:

1. check that the `kubectl port-forward` daemon command is indeed running on your system (e.g. by listing the processes with `ps -ef|grep kubectl`). You should see something like `kubectl --context gke_ai-gilde_europe-west4-a_cluster-1 --namespace kubeflow port-forward svc/metadata-grpc-service 8081:8080`

2. if the answer to 1. is yes, then try to access the gRPC port yourself with `curl`. If the port is open, you should get an answer that looks like the one below:

  ```
  $ curl -v http://localhost:8081/
  *   Trying 127.0.0.1:8081...
  * TCP_NODELAY set
  * Connected to localhost (127.0.0.1) port 8081 (#0)
  > GET / HTTP/1.1
  > Host: localhost:8088
  > User-Agent: curl/7.68.0
  > Accept: */*
  > 
  * Received HTTP/0.9 when not allowed
  
  * Closing connection 0
  curl: (1) Received HTTP/0.9 when not allowed
  ```
3. check the metadata UI daemon logs as indicated in the `zenml stack up` output (e.g. by looking at the /home/victor/.config/zenml/kubeflow/eda1b333-39a8-4a3b-9647-0f00fb78dab2/kubeflow_daemon.log file) and see if there is anything logged there that might point to a problem with the port-forwarding.
4. if all previous points check out, you can try connecting to the Kubeflow gRPC metadata service manually by running something like this (this is more or less what the `zenml stack up` command check is doing):
  ```python
  from zenml.repository import Repository
  r = Repository()
  r.get_pipelines()
  ```

Basically, the problem could be one of the following:
* the port is not forwarded locally (the `kubectl port-forward` command is failing)
* the gRPC metadata service and pods running in the GKE cluster are in a failure state
* there might be a problem with the metadata protocol in the particular version of Kubeflow that you have installed. You can consult the gRPC metadata pod logs to see if there is anything relevant logged there (e.g. `kubectl -n kubeflow logs metadata-grpc-deployment-6b5685488-48x4b`)

**Comment by stefannica** [comment]
Hey Victor, I think this happens because the pipeline is still running when the Tensorboard visualizer call is reached in the post-execution workflow, and the pipeline hasn't yet had a chance to be recorded in the metadata store.

There are two ways you can address this. You can either reconfigure your orchestrator to wait until the pipeline run is complete before it starts the visualizer (see below), or you can simply run the example a second time and the visualizer will find the previous pipeline run and start Tensorboard.

To have a synchronous type of pipeline execution in which your client code waits until the pipeline run is complete, you can run something like this:

```
zenml orchestrator update --synchronous=True
```

**Comment by htahir1** [comment]
@VictorW96 Can I close this issue in favor of #729 ?
